### PR TITLE
Create Passport Registration

### PR DESCRIPTION
**Pull request by dylanesque**
WIP, this was getting a little lengthy, I wanted to check in with current progress before adding invalid form state styling, DRYing up the CSS, and adding a snapshot test.

Edit: This latest version includes a fix for testing setup (it wasn't working previously) and a refactor to utilize Grid for the form layout.

Edit 2: Final revision? The one thing I'm unhappy with here is the button feels uncannily off from the center a bit. Widening the width of the grid container helped a little, but it looks "wrong" to me. I'm not sure how to fix this apart from removing the button and the link directly underneath it from out of the grid.

### CODE REVIEW

File: frontend/components/StateSelect/UsStateSelect.tsx
Comment (JMStudiosJoe) [inline frontend/components/StateSelect/UsStateSelect.tsx:5]: out of curiosity why not use a library for this?

File: frontend/pages/register/index.tsx
Comment (JMStudiosJoe) [inline frontend/pages/register/index.tsx:7]: I am pretty sure the flow is always the user must register as a basic user then from there once they are logged in they can then request to update to a passport user or behyond.

File: frontend/pages/register/passport/index.tsx
Comment (JMStudiosJoe) [inline frontend/pages/register/passport/index.tsx:25]: no need for form

File: frontend/pages/register/passport/passport.module.css
Comment (JMStudiosJoe) [inline frontend/pages/register/passport/passport.module.css:11]: I am not sure about the .button class naming but could we switch from css flex to css grid? its far more declarative and descriptive for what it is actually doing https://css-tricks.com/snippets/css/complete-guide-grid/

**Review by JMStudiosJoe — CHANGES_REQUESTED** [review]

File: frontend/pages/register/index.tsx
Comment (JMStudiosJoe) [reply frontend/pages/register/index.tsx:7]: This comment may be disregarded since i see the PR for basic user registration so maybe just need to link this one from the dashboard?

File: frontend/pages/register/passport/passport.module.css
Comment (dylanesque) [reply frontend/pages/register/passport/passport.module.css:11]: Yup, I can refactor that out, especially since I've seen the mobile mocks and have a clearer idea of what this should look like on all screens.

File: frontend/components/StateSelect/UsStateSelect.tsx
Comment (dylanesque) [reply frontend/components/StateSelect/UsStateSelect.tsx:5]: For the forms, or the state selection component specifically? At a glance, I'm unsure if pulling in a library for one component justifies the extra dependency we can avoid by just having this pasted in.

File: frontend/pages/register/index.tsx
Comment (dylanesque) [reply frontend/pages/register/index.tsx:7]: It's just a placeholder for me to test that routing was working as expected, it certainly doesn't need to stay here.

File: frontend/pages/register/passport/index.tsx
Comment (dylanesque) [reply frontend/pages/register/passport/index.tsx]: So, I don't agree with this. Even if it passes a11y checks on Lighthouse, using some other container is much worse from a semantic point of view. This component takes several pieces of input from the user and will (eventually) send it to the back-end, there is no better role for it than that of form.

Comment (JMStudiosJoe) [reply frontend/pages/register/passport/index.tsx]: That is all true however from my experience form element has introduced too many side effects and doesn't need to follow the semantic point of view for the risk of side effects involved. We have the component already maintaining the data entered in and do not need the HTML inputs and forms to rely on storing the data, so it fits another separation of concerns by keeping all the core logic in the JS and HTML only for structure.

I am willing to discuss further and we can bring this to the rest of the team as well and see their thoughts, but in my experience using the form element, even for the semantic benefit, just adds too much complexity for the simplicity of the component. Please let me know if you want to further discuss since I will need some convincing on this subject and always a great opportunity to div into something deeper for us to learn. :)

Comment (dylanesque) [reply frontend/pages/register/passport/index.tsx]: To be clear, is your objection to having a <form> tag in the codebase period, or are you assuming that I'm writing this with the attention of submitting the input data via that form element? Because if it's the latter, that's not my intention.

Comment (JMStudiosJoe) [reply frontend/pages/register/passport/index.tsx]: hmmm So then why use the form? I think my objective aligns with not using <form> at all

Comment (dylanesque) [reply frontend/pages/register/passport/index.tsx]: For semantic and a11y reasons, which going against creates more problems for us than it would solve.

Comment (JMStudiosJoe) [reply frontend/pages/register/passport/index.tsx]: oh to use <form> for accessibility reasons (a11y?) hmmm after doing some googling (https://gomakethings.com/why-use-a-form-element-when-submitting-fields-with-javascript/) looks like you are on the right track and they are more important than I thought. Just to clarify, will only be using for semantic purposes and for a11y reasons and not going to use the built in other stuff with it? In the past that is where I have run into too many issues with using a form but if just for semantic use and a11y bare minimum stuff then sounds good to me.

Comment (JMStudiosJoe) [reply frontend/pages/register/passport/index.tsx]: Great discussion!

Comment (dylanesque) [reply frontend/pages/register/passport/index.tsx]: Yeah, I'm not going to say I'll NEVER submit code that isn't idiomatic to the React way of doing things, but that's pretty high on my list of what I think about when writing. Submitting data via form elements is definitely on the list of "don't do this".

**Review by JMStudiosJoe — DISMISSED** [review]

File: frontend/components/StateSelect/UsStateSelect.tsx
Comment (srsexton94) [reply frontend/components/StateSelect/UsStateSelect.tsx:5]: Just un-resolving in case you missed my comment!

File: frontend/components/Logo/logo.module.css
Comment (srsexton94) [inline frontend/components/Logo/logo.module.css:3]: I think we're using `rem` over `px` - but my last merge introduced some variables you can use - `var(--size160)` will map to the right `rem` value for 160px!

File: frontend/pages/register/passport/index.tsx
Comment (srsexton94) [inline frontend/pages/register/passport/index.tsx:7]: Curious why this is outside of `Passport`!

**Review by srsexton94 — CHANGES_REQUESTED** [review]
I think this ticket has scope-crept pretty significantly - my understanding was this was a frontend ticket to create the UI of the Passport application page - but it seems only 4-7 of the 50 file changes have to do with that task

I think it'd probably be a good idea to break out the extra work you did into their own tickets/branches so that it's clear what we're trying to accomplish with each & so that the appropriate people can review your PR successfully

Screenshots/explanatory comments could also be helpful if it's not too much trouble!

File: frontend/components/Logo/logo.module.css
Comment (dylanesque) [reply frontend/components/Logo/logo.module.css:3]: Yeah, I was actually waiting for your branch to get merged so I could fold in any styling you added that overlapped, and also see what conventions you were using to hew closer to those.

File: frontend/components/StateSelect/UsStateSelect.tsx
Comment (dylanesque) [reply frontend/components/StateSelect/UsStateSelect.tsx:5]: Maybe? 

> I think this ticket has scope-crept pretty significantly - my understanding was this was a frontend ticket to create the UI of the Passport application page - but it seems only 4-7 of the 50 file changes have to do with that task
> 
> I think it'd probably be a good idea to break out the extra work you did into their own tickets/branches so that it's clear what we're trying to accomplish with each & so that the appropriate people can review your PR successfully
> 
> Screenshots/explanatory comments could also be helpful if it's not too much trouble!

Something is very wrong, there are a bunch of changes I didn't make, and I'm not sure why they're there. This will hopefully be more normalized once I pull in the latest changes from master.

File: frontend/pages/register/index.tsx
Comment (srsexton94) [inline frontend/pages/register/index.tsx]: Is this also outdated perhaps? I see in the thread above that this was meant to be a placeholder && it should've been replaced with `pages/viewer-registration/viewer-registration.tsx` with the last merge

File: frontend/components/StateSelect/select.module.css
Comment (srsexton94) [inline frontend/components/StateSelect/select.module.css:1]: Same comment as the other css file re: context-specific & user-facing naming!

File: frontend/components/Logo/Logo.tsx
Comment (srsexton94) [inline frontend/components/Logo/Logo.tsx:1]: We already have a Logo component at `shared-components/logo.tsx` (& it is/should be sized 144x113 as per wireframes)

**Review by srsexton94 — CHANGES_REQUESTED** [review]
Re-reviewed after changes - seems like there may have been some confusion (ie duplication/reverting) resolving merge conflicts? Happy to help resolve those if you like!

File: frontend/styles/globals.css
Comment (srsexton94) [reply frontend/styles/globals.css]: It does appear to be so! Apologies that "Outdated" flag wasn't on it when I was reviewing

File: frontend/pages/register/passport/passport.module.css
Comment (srsexton94) [reply frontend/pages/register/passport/passport.module.css:71]: Here's an article about it I found helpful!
https://engageinteractive.co.uk/blog/em-vs-rem-vs-px

File: frontend/pages/register/index.tsx
Comment (srsexton94) [reply frontend/pages/register/index.tsx]: Also appears to be so! Apologies, the "file changes" tab doesn't feel super intuitive for review

**Comment by dylanesque** [comment]
Regarding CSS classes: One of the nicer perks of CSS modules (that I wasn't aware of until this afternoon) is that the first part of the filename gets namespaced onto the class name in the browser, so these names are less sparse than they appear on paper.

**Comment by dylanesque** [comment]
@srsexton94 sr So, for the case of the state select component, I think making it a list to map through and render is more WET than anything else. Consider that:

- The list of states will never change from the back-end.
- The list will also never change from the front-end.

Given that, I don't see how adding JS to do a job that HTML is more than up to the task of helps us in this case.

File: frontend/shared-components/state-select/UsStateSelect.tsx
Comment (srsexton94) [inline frontend/shared-components/state-select/UsStateSelect.tsx:6]: I still think we should iterate through a string array of states to generate these more DRY-ly!

Something like `{ states.map(state => <option key={state} valued={state}>{state}</option>) }` in place of all the options (with `states: string[] = ['AL', 'AK', 'AR', etc....]`)

EDIT: I only just saw your response, but couldn't add a reply outside of a github "review" - but the above is much DRYer because you're not writing the option element 50 times over (literally 😅) 
It doesn't matter that the array won't change, it can be a constant - the point is just that it's just so repetitive

File: frontend/pages/passport-application/index.tsx
Comment (srsexton94) [inline frontend/pages/passport-application/index.tsx:36]: Hey! So thanks for removing the Bob Boberton stuff 😅 but the wireframes actually don't have any placeholder text at all & that's intentional as it's considered inaccessible/bad UX

Here's an article I found helpful on why that is! 
https://www.deque.com/blog/accessible-forms-the-problem-with-placeholders/

File: frontend/pages/passport-application/index.tsx
Comment (srsexton94) [inline frontend/pages/passport-application/index.tsx:13]: If you like, on other components we've been destructuring the classNames from the `styles` object (especially when there's more than like 1-3 classNames) 

It helps DRY up the template code a bit not having to write `styles.` repeatedly!

File: frontend/pages/register/index.tsx
Comment (srsexton94) [inline frontend/pages/register/index.tsx:1]: Was this potentially added back in re: merge conflicts? This file shouldn't exist anymore - it was replaced with `viewer-registration`

**Review by srsexton94 — CHANGES_REQUESTED** [review]
Just a final few comments! Left a response re: the states dropdown in the file

Also re: the classNames thing - even though CSS modules are name-spaced to the component my suggestion to use more informative names was more just for developer clarity - definitely not a blocker but I think there's a happy middle group to reached between specificity & brevity!

**Comment by JMStudiosJoe** [comment]
can we fix the conflicts @srsexton94 and then would it be good to be reviewed and merged?

**Comment by dylanesque** [comment]
Regarding the state component: Refactoring would be a lot more work since all I did was pull that HTML snippet off the web and plop it in the codebase: it doesn't get DRYer than "CMD + C, CMD + V, using working code for a common use case in forms". Changing that would be more work which makes the code there marginally less efficient since JavaScript is a render-blocking resource. Since the list of states never, ever changes, using that snippet makes sense because we're not trying to reinvent the wheel.

**Comment by srsexton94** [comment]
Hmm, I'm wondering if this state disagreement is coming from a misunderstanding of what DRY code means. [Here's a good/quick overview from one of my favorite clean-code gurus, Uncle Bob :)](https://youtu.be/7EmboKQH8lM?t=5828)

DRY doesn't have anything to do with  HOW you got the code (ie copy-pasted or original). It's about the code itself. The state solution you have is undeniably pretty repetitive code as It repeats the same code structure 50 times. 
I also feel strongly that falling back on "don't reinvent the wheel" (or its nefarious cousin "if it ain't broke don't fix it") can more often than not just act as an excuse to leave messy code (which I 100% know is not your intent!)

I of course didn't mean to make you feel like you had to do extra work though - to help relieve some of that workload here it is written in full - you should be able to just copy/paste this in!
```ts
const states: string[] = [
  "AL", "AK", "AR", "AZ", "CA", "CO", "CT", "DC", "DE", "FL", "GA", "HI", "IA", "ID", "IL", "IN", "KS", 
  "KY", "LA", "MA", "MD", "ME", "MI", "MN", "MO", "MS", "MT", "NC", "NE", "NH", "NJ", "NM", "NV", "NY", 
  "ND", "OH", "OK", "OR", "PA", "RI", "SC", "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WI", "WV", "WY"
]

export function USStateSelect() {
  return (
    <select className={styles.select}>
      { states.map(state => <option key={state} valued={state}>{state}</option>) }
    </select>
  )
}
```

File: frontend/README.md
Comment (srsexton94) [inline frontend/README.md:18]: Typo correction never fails to make my heart happy 👏🏻🥳🙌🏻

File: frontend/pages/passport-application/index.tsx
Comment (srsexton94) [inline frontend/pages/passport-application/index.tsx:88]: Apologies if not clear, this was part of the placeholder-text removal comment from before!

File: frontend/pages/passport-application/index.tsx
Comment (srsexton94) [inline frontend/pages/passport-application/index.tsx:56]: This differs fairly significantly from the error-validation patterns used in Login & ViewerRegistration - BUT imho the error validation stuff felt like it scope-crept wayyy too far out of a reasonable size for these mainly-UI-based tickets

Definitely not a blocker but I think we might want to create a follow-up ticket to work on error validation between the tree "enrollment" type pages!

File: frontend/pages/passport-application/index.tsx
Comment (srsexton94) [inline frontend/pages/passport-application/index.tsx:30]: Not a blocker, but if you wanted to reduce the length here you could utilize CSS nesting to not need quite so many classNames from the module!

**Review by srsexton94 — DISMISSED** [review]

**Comment by srsexton94** [comment]
Hey folks!  Mike & I talked offline - he's going to post some last changes & I've otherwise approved - please leave it for him to merge himself though to make sure everything's ready :)

**Review by dylanesque — COMMENTED** [review]
Merging and closing PR

> Hey folks! Mike & I talked offline - he's going to post some last changes & I've otherwise approved - please leave it for him to merge himself though to make sure everything's ready :)

Final changes are in, including a satisfactory fix for the button centering, this just needs approval from someone who isn't me :)

**Review by srsexton94 — APPROVED** [review]

**Pull request merged**
